### PR TITLE
[WIP] Add Optical Flow Driver for Upflow LC302 module

### LINKED ISSUE #13631: [Feature Request] Support UP-Flow LC302 Optical Flow Module
UP-Flow LC302 is a very popular optical flow module in china, lots of University are using this module for indoor usage. And toys also use this chip a lot.

This also have better performance than PMW3901.

image
image

This PCB maybe looks ugly, I'm working with them for next version.

Purchase: https://www.aliexpress.com/item/32968961948.html?spm=a2g0o.productlist.0.0.4db0495eSoA2ss&algo_pvid=aadf0f14-60af-47be-b95a-0ff85d514030&algo_expid=aadf0f14-60af-47be-b95a-0ff85d514030-0&btsid=627f14df-6157-45b0-9595-d32cdf348a09&ws_ab_test=searchweb0_0,searchweb201602_10,searchweb201603_55

### PR DESCRIPTION
**Pull request by bys1123**
Upflow LC302 has a very nice performance like PMW3901. Worth to upstream.

image

[Product User Manual(LC-302-3C)V1.1.pdf](https://github.com/PX4/Firmware/files/4513167/Product.User.Manual.LC-302-3C.V1.1.pdf)

This commit is an old one, need some work to make it compatible with current master.

This PR will fix #13631 .

### CODE REVIEW

**Comment by hamishwillee** [comment]
Great. When this is integrated can you update the record in http://docs.px4.io/master/en/sensor/pmw3901.html ?

**Comment by bys1123** [comment]
@hamishwillee Glad to do, but this one is not using PMW3901, and have its own protocol, is that a good idea? Should I create a new page? Or create 3901-like section is good to go?

**Comment by bys1123** [comment]
Just read the page again, I think create a new page is better.

File: src/drivers/optical_flow/upflow/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,44 @@
+############################################################################
+#
+#   Copyright (c) 2015 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/optical_flow/upflow/CMakeLists.txt:3]:
```suggestion
#   Copyright (c) 2020 PX4 Development Team. All rights reserved.
```

File: src/drivers/optical_flow/upflow/lc302.cpp [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+	upflow_lc302 *dev = reinterpret_cast<upflow_lc302 * >(arg);
+
+	if (dev->_fd != -1) {
+		dev->_cycle();
+
+	} else {
+		dev->_fd_open();
+	}
+
+	work_queue(HPWORK, &(dev->_work), (worker_t)&upflow_lc302::cycle_trampoline,
```
Comment (dagar) [inline src/drivers/optical_flow/upflow/lc302.cpp:447]: Needs to be updated to use the new WorkQueue.

File: src/drivers/optical_flow/CMakeLists.txt [hunk 3]
```diff
@@ -34,3 +34,4 @@
 add_subdirectory(paw3902)
 add_subdirectory(pmw3901)
 add_subdirectory(px4flow)
+add_subdirectory(upflow)
```
Comment (dagar) [inline src/drivers/optical_flow/CMakeLists.txt:37]: Should we name this more descriptively like `upflow_lc302`?

**Comment by dagar** [comment]
@bys1123 are you setup to flight test this? If so I can take a pass to clean it up.

File: src/drivers/optical_flow/CMakeLists.txt (hunk 3 above)
Comment (bys1123) [reply src/drivers/optical_flow/CMakeLists.txt:37]: I think If they have new module, they would continue use this protocol.

File: src/drivers/optical_flow/upflow/lc302.cpp (hunk 2 above)
Comment (bys1123) [reply src/drivers/optical_flow/upflow/lc302.cpp:447]: Yes, but I'm not familiar on this part, I need your help.

**Comment by bys1123** [comment]
@dagar I'm ready to test this, and the manufacture also will do the test too, I need your help to make this driver up-to-date, thank you.

**Comment by hamishwillee** [comment]
@bys1123 Yes please - create a new page for a new implementation that is unrelated to the others. I'll help with subedit.

File: src/drivers/optical_flow/upflow/lc302.cpp (hunk 2 above)
Comment (bys1123) [reply src/drivers/optical_flow/upflow/lc302.cpp]: Tried to fixed.

**Pull request closed without merging**